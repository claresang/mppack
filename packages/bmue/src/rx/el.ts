import { TPath, TRxLang, TRxLangHandler } from '../types';

export class ELang {
  constructor(name: string, lang: TRxLang) {
    this.name = name;
    this.lang = lang;
  }

  name: string;
  lang: TRxLang;

  parse() {
    const len = this.lang.length;
    const handler = this.lang[len - 1] as TRxLangHandler;
    const deps = this.lang.slice(0, len - 1) as Array<TPath>;
    return {
      name: this.name,
      deps,
      handler
    };
  }
}

//Factory method
export const EL = (name: string, lang: TRxLang) => new ELang(name, lang);
