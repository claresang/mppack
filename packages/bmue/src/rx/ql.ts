import { TPath, TQueryLang, TQueryLangHandler } from '../types';

let uuid = 0;

export class QueryLang {
  constructor(name: string, lang: TQueryLang) {
    this.id = ++uuid;
    this.name = name;
    this.lang = lang;
  }

  id: number;
  name: string;
  lang: TQueryLang;

  parse() {
    const len = this.lang.length;
    const handler = this.lang[len - 1] as TQueryLangHandler;
    const deps = this.lang.slice(0, len - 1) as Array<TPath>;
    return {
      id: this.id,
      name: this.name,
      deps,
      handler
    };
  }
}

//Factory method
export const QL = (name: string, lang: TQueryLang) => new QueryLang(name, lang);
