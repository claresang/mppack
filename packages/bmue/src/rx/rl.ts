import { TPath, TRxLang, TRxLangHandler } from '../types';

export class RxLang {
  constructor(name: string, lang: TRxLang) {
    this.name = name;
    this.lang = lang;
  }

  name: string;
  lang: TRxLang;

  parse() {
    const len = this.lang.length;
    const handler = this.lang[len - 1] as TRxLangHandler;
    const deps = this.lang.slice(0, len - 1) as Array<TPath>;
    return {
      name: this.name,
      deps,
      handler
    };
  }
}

//Factory method
export const RL = (name: string, lang: TRxLang) => new RxLang(name, lang);
