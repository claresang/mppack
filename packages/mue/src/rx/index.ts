import { getPathVal } from '../util';
import { QueryLang } from './ql';
import { RxLang } from './rl';

export default class RxFactroy {
  dev = false;
  data = {};
  _cacheQL = {} as { [key: number]: Array<any> };

  bigQuery(ql: QueryLang) {
    const { id, name, handler, deps } = ql.parse();

    if (this.dev) {
      console.groupCollapsed(
        `=================ql#${name}=====================`
      );
    }

    //initialize querylang cache
    !this._cacheQL[id] && (this._cacheQL[id] = []);

    let isChanged = false;
    deps.forEach((dep, i) => {
      //suport ql nested ql
      let val = null;
      if (dep instanceof QueryLang) {
        val = this.bigQuery(dep);
      } else {
        val = getPathVal(this.data, dep);
      }
      if (val != this._cacheQL[id][i]) {
        isChanged = true;
        this._cacheQL[id][i] = val;
      }
    });

    if (isChanged) {
      const len = deps.length;
      const pathVals = this._cacheQL[id].slice(0, len);
      const result = handler(...pathVals);
      this._cacheQL[id][len + 1] = result;

      if (this.dev) {
        console.log('update:ðŸ”¥', result);
        console.groupEnd();
      }

      return result;
    } else {
      const len = deps.length;
      const result = this._cacheQL[id][len + 1];

      if (this.dev) {
        console.log('cache:ðŸ˜†', result);
        console.groupEnd();
      }

      return result;
    }
  }

  _parseRL(rl: RxLang) {
    const cache = [];
    const { name, deps, handler } = rl.parse();
    if (this.dev) {
      console.groupCollapsed(
        `=================rl#${name}=====================`
      );
    }
    for (let dep of deps) {
      cache.push(
        dep instanceof QueryLang
          ? this.bigQuery(dep)
          : getPathVal(this.data, dep)
      );
    }

    return () => {
      let isChanged = false;
      deps.forEach((dep, i) => {
        let pathVal =
          dep instanceof QueryLang
            ? this.bigQuery(dep)
            : getPathVal(this.data, dep);
        if (pathVal != cache[i]) {
          cache[i] = pathVal;
          isChanged = true;
        }
      });

      if (isChanged) {
        if (this.dev) {
          console.log('changed:ðŸ”¥');
          console.groupEnd();
        }
        handler(...cache);
      } else {
        if (this.dev) {
          console.log('cache:ðŸ˜†');
          console.groupEnd();
        }
      }
    };
  }
}
