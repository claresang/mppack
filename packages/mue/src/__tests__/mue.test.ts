import Mue from '../mue';
import { QL } from '../rx/ql';
import { RL } from '../rx/rl';

//mock miniapp Page object
window['Page'] = jest.fn(obj => obj);

it('test mue', () => {
  const mue = Mue({
    data: {
      list: [{ id: 1, name: 'mue', mott: 'easy miniapp' }]
    },
    action: {
      sayHello() {}
    },
    getter: {
      hello: QL('helloQL', [
        // data path
        ['list', 0],
        ({ name, mott }) => `${name}-${mott}`
      ])
    },
    effect: [
      RL('helloRL', [
        //data path
        ['list', 0, 'name'],
        name => console.log(name)
      ])
    ]
  });
  expect(mue).toMatchSnapshot();
});
