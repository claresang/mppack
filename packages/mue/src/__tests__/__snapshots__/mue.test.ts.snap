// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test mue 1`] = `
Object {
  "_cacheQL": Object {},
  "data": Object {},
  "dev": false,
  "effect": Array [
    RxLang {
      "lang": Array [
        Array [
          "list",
          0,
          "name",
        ],
        [Function],
      ],
      "name": "helloRL",
    },
  ],
  "getter": Object {
    "hello": QueryLang {
      "id": 1,
      "lang": Array [
        Array [
          "list",
          0,
        ],
        [Function],
      ],
      "name": "helloQL",
    },
  },
  "sayHello": [Function],
}
`;
